// @figma-export/core/dist/lib/figma.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.enrichPagesWithSvg = exports.fileSvgs = exports.getImages = exports.getClient = exports.getIdsFromPages = exports.getPages = exports.getComponents = void 0;
const Figma = __importStar(require("figma-js"));
const path_1 = require("path");
const p_limit_1 = __importDefault(require("p-limit"));
const p_retry_1 = __importDefault(require("p-retry"));
const utils_1 = require("./utils");
const getComponents = (children = [], filter = () => true) => {
    let components = [];
    children.forEach((component) => {
        if (component.type === 'COMPONENT' && filter(component)) {
            components.push(Object.assign(Object.assign({}, component), { svg: '', figmaExport: {
                    id: component.id,
                    dirname: (0, path_1.dirname)(component.name),
                    basename: (0, path_1.basename)(component.name),
                } }));
            return;
        }
        if ('children' in component) {
            components = [
                ...components,
                ...getComponents((component.children), filter),
            ];
        }
    });
    return components;
};
exports.getComponents = getComponents;
const filterPagesByName = (pages, pageNames = []) => {
    const only = (0, utils_1.toArray)(pageNames).filter((p) => p.length);
    return pages.filter((page) => only.length === 0 || only.includes(page.name));
};
const getPages = (document, options = {}) => {
    const pages = filterPagesByName(document.children, options.only);
    return pages
        .map((page) => (Object.assign(Object.assign({}, page), { components: getComponents(page.children, options.filter) })))
        .filter((page) => page.components.length > 0);
};
exports.getPages = getPages;
const getIdsFromPages = (pages) => pages.reduce((ids, page) => [
    ...ids,
    ...page.components.map((component) => component.id),
], []);
exports.getIdsFromPages = getIdsFromPages;
const getClient = (token) => {
    if (!token) {
        throw new Error('\'Access Token\' is missing. https://www.figma.com/developers/docs#authentication');
    }
    return Figma.Client({ personalAccessToken: token });
};
exports.getClient = getClient;
const fileImages = (client, fileId, ids) => __awaiter(void 0, void 0, void 0, function* () {
    const response = yield client.fileImages(fileId, {
        ids,
        format: 'svg',
        svg_include_id: true,
    }).catch((error) => {
        throw new Error(`while fetching fileImages: ${error.message}`);
    });
    return response.data.images;
});
const getImages = (client, fileId, ids) => __awaiter(void 0, void 0, void 0, function* () {
    const idss = (0, utils_1.chunk)(ids, 200);
    const limit = (0, p_limit_1.default)(30);
    const resolves = yield Promise.all(idss.map((groupIds) => {
        return limit(() => fileImages(client, fileId, groupIds));
    }));
    return Object.assign({}, ...resolves);
});
exports.getImages = getImages;
const fileSvgs = (client, fileId, ids, { concurrency = 30, retries = 3, transformers = [], 
// eslint-disable-next-line @typescript-eslint/no-empty-function
onFetchCompleted = () => { }, } = {}) => __awaiter(void 0, void 0, void 0, function* () {
    const images = yield getImages(client, fileId, ids);
    const limit = (0, p_limit_1.default)(concurrency);
    let index = 0;
    const svgPromises = Object.entries(images).map(([id, url]) => __awaiter(void 0, void 0, void 0, function* () {
        const svg = yield limit(() => (0, p_retry_1.default)(() => (0, utils_1.fetchAsSvgXml)(url), { retries }));
        const svgTransformed = yield (0, utils_1.promiseSequentially)(transformers, svg);
        onFetchCompleted({
            index: index += 1,
            total: ids.length,
        });
        return [id, svgTransformed];
    }));
    const svgs = yield Promise.all(svgPromises);
    return (0, utils_1.fromEntries)(svgs);
});
exports.fileSvgs = fileSvgs;
const enrichPagesWithSvg = (client, fileId, pages, svgOptions) => __awaiter(void 0, void 0, void 0, function* () {
    const componentIds = getIdsFromPages(pages);
    if (componentIds.length === 0) {
        throw new Error('No components found');
    }
    const svgs = yield fileSvgs(client, fileId, componentIds, svgOptions);
    return pages.map((page) => (Object.assign(Object.assign({}, page), { components: page.components.map((component) => (Object.assign(Object.assign({}, component), { svg: svgs[component.id] || utils_1.emptySvg }))) })));
});
exports.enrichPagesWithSvg = enrichPagesWithSvg;
