// @figma-export/core/dist/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.notEmpty = exports.fetchAsSvgXml = exports.chunk = exports.promiseSequentially = exports.fromEntries = exports.emptySvg = exports.toArray = void 0;
const axios_1 = __importDefault(require("axios"));
const toArray = (any) => (Array.isArray(any) ? any : [any]);
exports.toArray = toArray;
exports.emptySvg = '<svg></svg>';
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const fromEntries = (iterable) => {
    return [...iterable].reduce((obj, [key, val]) => {
        // eslint-disable-next-line no-param-reassign
        obj[key] = val;
        return obj;
    }, {});
};
exports.fromEntries = fromEntries;
// eslint-disable-next-line @typescript-eslint/ban-types
const promiseSequentially = (promiseFactories, initialValue) => {
    const promise = promiseFactories.reduce((previousPromise, promiseFactory) => {
        return previousPromise.then((value) => promiseFactory(value));
    }, Promise.resolve(initialValue));
    return promise;
};
exports.promiseSequentially = promiseSequentially;
const chunk = (array, perChunk) => {
    return array.reduce((all, one, i) => {
        const ch = Math.floor(i / perChunk);
        // eslint-disable-next-line no-param-reassign
        all[ch] = [...(all[ch] || []), one];
        return all;
    }, []);
};
exports.chunk = chunk;
const fetchAsSvgXml = (url) => {
    if (!/https?:\/\/.*/.test(url)) {
        throw new TypeError('Only absolute URLs are supported');
    }
    return axios_1.default.get(url, {
        headers: {
            'Content-Type': 'images/svg+xml',
        },
    }).then((response) => {
        if (response.data.length === 0)
            return exports.emptySvg;
        return response.data;
    }).catch((error) => {
        throw new Error(`while fetching svg "${url}": ${error.message}`);
    });
};
exports.fetchAsSvgXml = fetchAsSvgXml;
const notEmpty = (value) => {
    return value !== null && value !== undefined;
};
exports.notEmpty = notEmpty;
