// @figma-export/cli/dist/commands/styles.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addStyles = void 0;
const figmaExport = __importStar(require("@figma-export/core"));
const utils_1 = require("../utils");
const addStyles = (prog, spinner) => prog
    .command('styles <fileId>')
    .describe('Export styles from a Figma file.')
    .option('-O, --outputter', 'Outputter module or path')
    .option('-o, --output', 'Output directory', 'output')
    .option('-p, --page', 'Figma page names (all pages when not specified)')
    .option('--fileVersion', `A specific version ID to get. Omitting this will get the current version of the file.
                         https://help.figma.com/hc/en-us/articles/360038006754-View-a-file-s-version-history`)
    .example('styles fzYhvQpqwhZDUImRz431Qo -O @figma-export/output-styles-as-css')
    .action((fileId, _a) => {
    var { fileVersion, output } = _a, opts = __rest(_a, ["fileVersion", "output"]);
    const outputter = (0, utils_1.asArray)(opts.outputter);
    const page = (0, utils_1.asArray)(opts.page);
    spinner.info(`Exporting ${fileId} as [${outputter.join(', ')}]`);
    spinner.start();
    figmaExport.styles({
        fileId,
        version: fileVersion,
        token: process.env.FIGMA_TOKEN || '',
        onlyFromPages: page,
        outputters: (0, utils_1.requirePackages)(outputter, { output }),
        // eslint-disable-next-line no-param-reassign
        log: (message) => { spinner.text = message; },
    }).then(() => {
        spinner.succeed('done');
    }).catch((error) => {
        spinner.fail();
        // eslint-disable-next-line no-console
        console.log(error);
    });
});
exports.addStyles = addStyles;
